compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -739,111 +739,117 @@ private STNode parseImportPrefix() {
+                reportInvalidQualifierList(qualifiers);
+                return parseEnumDeclaration(metadata, publicQualifier);
             case IDENTIFIER_TOKEN:
                 // Here we assume that after recovering, we'll never reach here.
                 // Otherwise the tokenOffset will not be 1.
                 if (isModuleVarDeclStart(1)) {
-                    return parseModuleVarDecl(metadata, qualifier);
+                    reportInvalidQualifierList(qualifiers);
+                    return parseModuleVarDecl(metadata, publicQualifier);
                 }
                 // fall through
             default:
+                if (isTopLevelQualifier(nextToken.kind)) {
+                    // Top level qualifiers are present. Parse them first and then try to re-parse the method.
+                    parseTopLevelQualifiers(qualifiers);
+                    return parseTopLevelNode(metadata, publicQualifier, qualifiers);
+                }
+
                 if (isTypeStartingToken(nextToken.kind) && nextToken.kind != SyntaxKind.IDENTIFIER_TOKEN) {
-                    return parseModuleVarDecl(metadata, qualifier);
+                    return parseModuleVarDeclWithoutQuals(metadata, publicQualifier, qualifiers);
                 }
 
                 STToken token = peek();
-                Solution solution =
-                        recover(token, ParserRuleContext.TOP_LEVEL_NODE_WITHOUT_MODIFIER, metadata, qualifier);
+                Solution solution = recover(token, ParserRuleContext.TOP_LEVEL_NODE_WITHOUT_MODIFIER, metadata,
+                        publicQualifier, qualifiers);
 
                 if (solution.action == Action.KEEP) {
                     // If the solution is {@link Action#KEEP}, that means next immediate token is
                     // at the correct place, but some token after that is not. There only one such
                     // cases here, which is the `case IDENTIFIER_TOKEN`. So accept it, and continue.
-                    return parseModuleVarDecl(metadata, qualifier);
+                    reportInvalidQualifierList(qualifiers);
+                    return parseModuleVarDecl(metadata, publicQualifier);
                 }
 
-                return parseTopLevelNode(metadata, qualifier);
+                return parseTopLevelNode(metadata, publicQualifier, qualifiers);
         }
-
     }
 
-    private boolean isObjectTypeStart() {
-        return peek(2).kind == SyntaxKind.OBJECT_KEYWORD || peek(3).kind == SyntaxKind.OBJECT_KEYWORD;
+    private STNode parseModuleVarDecl(STNode metadata, STNode qualifier) {
+        List<STNode> topLevelQualifiers = new ArrayList<>();
+        return parseModuleVarDeclWithoutQuals(metadata, qualifier, topLevelQualifiers);
     }
 
-    private STNode parseModuleVarDecl(STNode metadata, STNode qualifier) {
+    private STNode parseModuleVarDeclWithoutQuals(STNode metadata, STNode qualifier, List<STNode> qualifiers) {
```

Comment:
Its fine to call this `parseModuleVarDecl`. The overloaded parameters will indicate this takes qualifiers from outside..